proposed/runtime-binding.md

```diff
@@ -229,37 +225,80 @@ Can be specified via the following ways:
+
+The following policy will be used in 3.0:
+
+* If `ROLL_FORWARD_ON_NO_CANDIDATE_FX` is set, it will be honored.
+* If `ROLL_FORWARD` is set, it will be honored.
+* If both settings are set, it is an error.
+* If neither settings are set, then the existing default behavior is used, as described in the **Patch Version Selection** and **Patch Version Selection** sections above.
+
+Note: The ENV syntax is used above, but the same rules apply for any way that the two settings are set.
+
 ### fx-version
 
-You can specify the exact desired runtime version using the command line argument '--fx-version'. In this case, only the specified version will be accepted, even if patch roll forward is enabled. The expected behavior would be the same in the example above.
+You can specify the exact desired runtime version using the command line argument '--fx-version'.
 
-The following example shows how to use this knob:
+The following example shows the existing behavior of this knob:
 
 ```console
-C:\testapps\twotwoapp>dotnet --fx-version 3.0.0 bin\Debug\netcoreapp2.2\twotwoapp.dll
+C:\testapps\twooneapp>type Program.cs
+using System;
+
+namespace twooneapp
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            Console.WriteLine("Hello World!");
+            Console.WriteLine(typeof(Object).Assembly.Location);
+        }
+    }
+}
+C:\testapps\twooneapp>
+C:\testapps\twooneapp>dotnet bin\Debug\netcoreapp2.1\twooneapp.dll
+Hello World!
+C:\Program Files\dotnet\shared\Microsoft.NETCore.App\2.1.7\System.Private.CoreLib.dll
+
+C:\testapps\twooneapp>dotnet --fx-version 2.1.0 bin\Debug\netcoreapp2.1\twooneapp.dll
+Hello World!
+C:\Program Files\dotnet\shared\Microsoft.NETCore.App\2.1.0\System.Private.CoreLib.dll
 ```
 
-The example above will attempt to load the 3.0.0 runtime version. If it isn't found, the host will print an error message.
+The `--fx-version` argument can be composed with the new `RollForward` settings, either via CLI or ENV. By default, `--fx-version` does not roll-forward.
 
-Another example:
+The following example enables a .NET Core 2.1 application to run on .NET Core 2.2 without having to select the correct 2.2 patch version. In this case `--fx-version` specifies the floor version. In this scenario, 2.2.0 is not on the machine, but a later patch version is.
 
 ```console
-C:\testapps\threezeroapp>bin\Debug\netcoreapp3.0\threezero.exe --fx-version 3.0.0 --roll-forward patch
+C:\testapps\twooneapp>dotnet --fx-version 2.2.0 bin\Debug\netcoreapp2.1\twooneapp.dll
+It was not possible to find any compatible framework version
+The specified framework 'Microsoft.NETCore.App', version '2.2.0' was not found.
+C:\testapps\twooneapp>dotnet --fx-version 2.2.0 --roll-forward Patch bin\Debug\netcoreapp2.1\twooneapp.dll
+Hello World!
+C:\Program Files\dotnet\shared\Microsoft.NETCore.App\2.2.3\System.Private.CoreLib.dll
```

Comment:
I like the overall outcome of this. The only confusion might arise when all the different ways to specify these settings is combined. If specified on command line together like this, it makes perfect sense. If the `RollFoward` is specified through an environment variable which is effectively invisible, it gets a bit more complicated (as compared to existing behavior of `--fx-version`).
Basically if I specify `--fx-version` today I get some relatively predictable behavior. After this change that behavior becomes less predictable since the ambient env. variable may modify it. Now I have to know about the env. variable (and the `runtimeconfig` setting to be able to decide the exact behavior.

I'm not sure if it's OK to leave it like that, or if there's something we should do... and I don't know what... just raising a concern.

Reply:
I would have thought the order precedence would have been:

1. CLI arguments // This is the app user who knows the running the environment the best and is choosing to opt in newer behavior.
2. `.runtimeconfig.json` properties (AKA "json") // The developer believes he is using low level functionality that won't change much and is willing to take the risk it will work in the future but doesn't really not for sure.
3. Environment variables (AKA "ENV") // Useful when you don't control the how the process starts.

At AWS I can't imagine us ever setting the Environment variables on behalf of a customer because we can't guarantee the app would work. But we would want the developer running on AWS the option to override the default behavior by setting the environment variable.

For example a developer has deployed a .NET Core 2.1 Lambda Function (Serverless). Then we add .NET Core 3.X support. Currently the user would have to recompile the app to move it to the 3.X runtime. If the environment variable exists they could just modify the environment variable and change the runtime to be 3.X with no recompilation required. They just need to retest to make sure it works in 3.X.

Reply:
I see that I missed the key detail above `(later scopes take precedence over earlier ones)` and cli arguments will override anything in `.runtime.config` or environment variables.

For environment variables I think they should have the least amount of power to change the behavior given their potential large scope. If somebody explicitly made a setting in `.runtime.config` or command line arguments, those should be chosen over environment variables since the variable might not have been set for that particular application.

This logic is based on the comment earliar that tooling would not automatically set any the project settings so that a developer always has to explicitly make the choice to opt into their behavior.